internal/sqlparser/parser.go

```diff
@@ -277,6 +286,65 @@ func ParseSQLMigration(r io.Reader, direction Direction, debug bool) (stmts []st
 	return stmts, useTx, nil
 }
 
+type annotation string
+
+const (
+	annotationUp             annotation = "Up"
+	annotationDown           annotation = "Down"
+	annotationStatementBegin annotation = "StatementBegin"
+	annotationStatementEnd   annotation = "StatementEnd"
+	annotationNoTransaction  annotation = "NO TRANSACTION"
+	annotationEnvsubOn       annotation = "ENVSUB ON"
+	annotationEnvsubOff      annotation = "ENVSUB OFF"
+)
+
+var supportedAnnotations = map[annotation]struct{}{
+	annotationUp:             {},
+	annotationDown:           {},
+	annotationStatementBegin: {},
+	annotationStatementEnd:   {},
+	annotationNoTransaction:  {},
+	annotationEnvsubOn:       {},
+	annotationEnvsubOff:      {},
+}
+
+var (
+	errEmptyAnnotation   = errors.New("empty annotation")
+	errInvalidAnnotation = errors.New("invalid annotation")
+)
+
+// extractAnnotation extracts the annotation from the line.
+// All annotations must be in format: "-- +goose [annotation]"
+// Allowed annotations: Up, Down, StatementBegin, StatementEnd, NO TRANSACTION, ENVSUB ON, ENVSUB OFF
+func extractAnnotation(line string) (annotation, error) {
+	// Remove leading and trailing whitespace from line.
+	cmd := strings.TrimSpace(line)
```

Comment:
I've been going back and forth on whether we should do this.

I.e., a goose annotation must begin with `-- +goose`, with no leading white space.

Reply:
I can remove this line and add a check if the line has a leading space and throw an appropriate error.

Reply:
That's probably the right thing:

1. Only allow `-- +goose ...` annotations without leading space
2. Raise an error if we detect a `-- +goose` annotation with leading whitespace like `\s+-- +goose`

Reply:
I'll add changes to cover this logic

Reply:
@mfridman I refactored PR to allow annotations only without leading whitespace